File: 0502-ipo/0502-ipo.java
class Solution {
    public int findMaximizedCapital(int k, int w, int[] profits, int[] capital) {
        int n = profits.length;
        List<Pair<Integer, Integer>> list = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            list.add(new Pair<>(capital[i], profits[i]));
        }
        list.sort((a, b) -> a.getKey().equals(b.getKey()) ? b.getValue() - a.getValue() : a.getKey() - b.getKey());
        int i = 0;
        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b - a);
        while (k-- > 0) {
            while (i < n && list.get(i).getKey() <= w) {
                pq.add(list.get(i).getValue());
                i++;
            }
            if (pq.isEmpty()) {
                break;
            }
            w += pq.poll();
        }
        return w;
    }
}

File: 0502-ipo/0502-ipo.cpp
class Solution {
public:
    int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
        int n = profits.size();
        vector<pair<int, int>> vec(n);
        for (int i = 0; i < n; i++) {
            vec[i] = {capital[i], profits[i]};
        }
        sort(vec.begin(), vec.end());
        int i = 0;
        priority_queue<int> pq;
        while (k--) {
            while (i < n && vec[i].first <= w) {
                pq.push(vec[i].second);
                i++;
            }
            if (pq.empty()) {
                break;
            }
            w += pq.top();
            pq.pop();
        }
        return w;
    }
};



// heap-buffer-overflow
// class Solution {
// public:
//     int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
//         vector<pair<int, int>> vec(n);
//         int n = profits.size();
//         for (int i = 0; i < n; i++) {
//             vec[i[ = {capital[i], profits[i]};
//         }

//         sort(vec.begin(), vec.end(), [](pair<int, int> &a, pair<int, int> &b) {
//             if (a.first == b.first) {
//                 return a.second > b.second;
//             }
//             return a.first < b.first;
//         });
        
//         int totalProfit = 0;
//         for (int i = 0; i < k; i++) {
//             pair<int, int> temp = vec[i];
//             if (temp.first <= w) {
//                 w += temp.second;
//             } else {
//                 break;
//             }
//         }
//         return w;
//     }
// };